rogue/src/entity_spawners/mod.rs

```diff
@@ -226,69 +205,140 @@ struct MonsterSpawnParameters {
+    spawn_table.insert(MonsterType::Bat,               MonsterSpawnParameters::new(1, 1, 4, 25, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::Snake,             MonsterSpawnParameters::new(1, 1, 4, 500, MonsterSpawnBound::Grass));
+    spawn_table.insert(MonsterType::GoblinBasic,       MonsterSpawnParameters::new(2, 1, 6, 25, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::GoblinCleric,      MonsterSpawnParameters::new(2, 2, 8, 20, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::GoblinEnchanter,   MonsterSpawnParameters::new(2, 2, 8, 20, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::GoblinFirecaster,  MonsterSpawnParameters::new(4, 3, 8, 15, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::GoblinChillcaster, MonsterSpawnParameters::new(4, 3, 8, 15, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::PinkJelly,         MonsterSpawnParameters::new(3, 4, 10, 15, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::OrangeJelly,       MonsterSpawnParameters::new(4, 4, 12, 10, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::BlueJelly,         MonsterSpawnParameters::new(4, 4, 12, 10, MonsterSpawnBound::None));
+    spawn_table.insert(MonsterType::Orc,               MonsterSpawnParameters::new(4, 5, 12, 25, MonsterSpawnBound::None));
     spawn_table
 }
 
+pub fn spawn_monsters(ecs: &mut World, depth: i32) {
+    let monsters_with_locations = sample_monster_spawn_vector(ecs, depth);
+    for (monster, (x, y)) in monsters_with_locations.into_iter() {
+        insert_monster(ecs, monster, x, y);
+    }
+}
+
+fn sample_monster_spawn_vector(ecs: &mut World, depth: i32) -> Vec<(MonsterType, (i32, i32))> {
+
+    let spawn_table = get_monster_spawn_table();
+    let mut rtable: random_table::RandomTable<(MonsterType, MonsterSpawnBound)>
+        = random_table::RandomTable::new();
+    for (monster_type, spawn_parameters) in spawn_table.iter() {
+        let ok_to_spawn_by_depth = (spawn_parameters.min_depth <= depth)
+            && (depth <= spawn_parameters.max_depth);
+        let ok_to_spawn_by_bound = match spawn_parameters.bound {
+            MonsterSpawnBound::None => true,
+            MonsterSpawnBound::Grass => {
+                let map = ecs.fetch::<Map>();
+                map.grass.iter().filter(|b| **b).count() > 0
+            }
+        };
+        if ok_to_spawn_by_depth && ok_to_spawn_by_bound {
+            rtable = rtable.insert((*monster_type, spawn_parameters.bound), spawn_parameters.chance);
+        }
+    }
+
+    let mut monsters_with_bounds: Vec<(MonsterType, MonsterSpawnBound)> = Vec::new();
+    let difficulty_quota: i32 = 10 + (5 * depth) / 2;
+    let mut seen_difficulty: i32 = 0;
+    let mut rng = ecs.write_resource::<RandomNumberGenerator>();
+    loop {
+        let (monster, bound) = rtable.roll(&mut rng).expect("Monster spawn sampling failed.");
```

Comment:
While reviewing the map connectedness PR the other day, one of the debugging flags I added was a flag to make it so you could descend from anywhere - to make it easier to generate/review a bunch of maps in quick sequence.

When I did this, I found that it was possible to get a panic out of the game by descending about a dozen floors. I added that same flag and tried it out, and got a panic out of here because the monster spawn table is empty. I assume that's at least semi-intentional, since none of the spawn parameters have a max depth >12. Is there a longer term plan for scaling, or otherwise ending the run?

Reply:
Absolutely no such plan 😆 

It would be fun to think about difficulty scaling, but there's also not many deeper dungeon monsters added to the game yet. I dunno, any thoughts?